src/main/java/ch/njol/skript/effects/EffSort.java

```diff
@@ -0,0 +1,167 @@
+})
+@Since("INSERT VERSION")
+@Keywords("input")
+public class EffSort extends Effect implements InputSource {
+
+	static {
+		Skript.registerEffect(EffSort.class, "sort %~objects% [(by|based on) <.+>]");
+		if (!ParserInstance.isRegistered(InputData.class))
+			ParserInstance.registerData(InputData.class, InputData::new);
+	}
+
+	@Nullable
+	private Expression<?> mappingExpr;
+	@Nullable
+	private String unparsedExpression;
+	private Variable<?> unsortedObjects;
+
+	private Set<ExprInput<?>> dependentInputs = new HashSet<>();
+
+	@Nullable
+	private Object currentValue;
+	@Nullable
+	private String currentIndex;
+
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (expressions[0].isSingle() || !(expressions[0] instanceof Variable)) {
+			Skript.error("You can only sort list variables!");
+			return false;
+		}
+		unsortedObjects = (Variable<?>) expressions[0];
+
+		if (parseResult.regexes.size() > 0) {
+			unparsedExpression = parseResult.regexes.get(0).group();
+			assert unparsedExpression != null;
+			InputData inputData = getParser().getData(InputData.class);
+			InputSource originalSource = inputData.getSource();
+			inputData.setSource(this);
+			mappingExpr = new SkriptParser(unparsedExpression, SkriptParser.PARSE_EXPRESSIONS, ParseContext.DEFAULT)
+				.parseExpression(Object.class);
+			inputData.setSource(originalSource);
+			return mappingExpr != null && mappingExpr.isSingle();
+		}
+		return true;
+	}
+
+	@Override
+	protected void execute(Event event) {
+		Object[] sorted;
+		if (mappingExpr == null) {
+			try {
+				sorted = unsortedObjects.stream(event)
+					.sorted(ExprSortedList::compare)
+					.toArray();
+			} catch (IllegalArgumentException | ClassCastException e) {
+				return;
+			}
+		} else {
+			Map<Object, Object> valueToMappedValue = new LinkedHashMap<>();
+			for (Iterator<Pair<String, Object>> it = unsortedObjects.variablesIterator(event); it.hasNext(); ) {
+				Pair<String, Object> pair = it.next();
+				currentIndex = pair.getKey();
+				currentValue = pair.getValue();
+				Object mappedValue = mappingExpr.getSingle(event);
+				if (mappedValue == null)
+					return;
+				valueToMappedValue.put(currentValue, mappedValue);
+			}
+			try {
+				sorted = valueToMappedValue.entrySet().stream()
+					.sorted(Map.Entry.comparingByValue(ExprSortedList::compare))
+					.map(Map.Entry::getKey)
+					.toArray();
+			} catch (IllegalArgumentException | ClassCastException e) {
+				return;
+			}
+		}
+
+		unsortedObjects.change(event, sorted, ChangeMode.SET);
+	}
+
+	@Override
+	public Set<ExprInput<?>> getDependentInputs() {
+		return dependentInputs;
+	}
+
+	@Override
+	public @Nullable Object getCurrentValue() {
+		return currentValue;
+	}
+
+	@Override
+	public boolean hasIndices() {
+		return true;
+	}
+
+	@Override
+	public String getCurrentIndex() {
+		assert currentIndex != null;
```

Comment:
isnt this assertion untrue

Reply:
You're right, it's a relic of when I was considering a different approach